server/src/main/java/io/spine/server/delivery/SystemEventWatcher.java

```diff
@@ -47,24 +53,48 @@
 
     private final TypeUrl targetType;
 
+    /**
+     * Crates a new instance of {@code SystemEventWatcher}.
+     *
+     * @param targetType
+     *         the type of the entity to watch the events for
+     */
     protected SystemEventWatcher(TypeUrl targetType) {
         super();
         this.targetType = checkNotNull(targetType);
     }
 
+    @Override
+    public Set<String> dispatch(EventEnvelope envelope) {
+        return producerOfCorrectType(envelope)
+               ? super.dispatch(envelope)
+               : emptySet();
+    }
+
     /**
-     * Tells whether or not the given {@link io.spine.system.server.EntityHistoryId EntityHistoryId}
-     * belongs to an entity of the {@link #targetType}.
+     * Checks if the given system event belongs to an entity history of an entity with
+     * the {@link #targetType}.
      *
-     * @param historyId
-     *         the entity history ID to check
-     * @return {@code true} if the type of the entity with given ID is the target type of this
-     *         subscriber, {@code false} otherwise
+     * @param event
+     *         the event to check
+     * @return {@code true} if the given event producer ID is an {@link EntityHistoryId} and
+     *         the {@linkplain EntityHistoryId#getTypeUrl() domain entity type} is the target type
+     *         of this watcher
      */
-    protected boolean isCorrectType(EntityHistoryId historyId) {
-        String typeUrlRaw = historyId.getTypeUrl();
-        TypeUrl typeUrl = TypeUrl.parse(typeUrlRaw);
-        return typeUrl.equals(targetType);
+    private boolean producerOfCorrectType(EventEnvelope event) {
+        EventContext context = event.getEventContext();
+        Any anyId = context.getProducerId();
+        boolean idTypeMatches = anyId.is(EntityHistoryId.class);
```

Comment:
The whole method and this expression in particular seem an expensive operation.

Let's discuss.